### PR TITLE
[Clang] Disable use of the counted_by attribute for whole struct pointers

### PR DESCRIPTION
**Pull request by bwendling**
The whole struct is specificed in the __bdos. The calculation of the whole size of the structure can be done in two ways:

    1) sizeof(struct S) + count * sizeof(typeof(fam))
    2) offsetof(struct S, fam) + count * sizeof(typeof(fam))

The first will add any remaining whitespace that might exist after allocation while the second method is more precise, but not quite expected from programmers. See [1] for a discussion of the topic.

GCC isn't (currently) able to calculate __bdos on a pointer to the whole structure. Therefore, because of the above issue, we'll choose to match what GCC does for consistency's sake.

[1] https://lore.kernel.org/lkml/ZvV6X5FPBBW7CO1f@archlinux/

### CODE REVIEW

**Review by Cydox — APPROVED** [review]
LGTM

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -1013,6 +1013,24 @@ CodeGenFunction::emitFlexibleArrayMemberSize(const Expr *E, unsigned Type,
...
     return nullptr;
 
+  if (isa<DeclRefExpr>(Base))
+    // The whole struct is specificed in the __bdos. The calculation of the
+    // whole size of the structure can be done in two ways:
+    //
+    //     1) sizeof(struct S) + count * sizeof(typeof(fam))
+    //     2) offsetof(struct S, fam) + count * sizeof(typeof(fam))
+    //
+    // The first will add any remaining whitespace that might exist after
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/CGBuiltin.cpp:1023]:
```suggestion
    // The first will add additional padding after the end of the array,
```

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -1013,6 +1013,24 @@ CodeGenFunction::emitFlexibleArrayMemberSize(const Expr *E, unsigned Type,
     // Can't find the field referenced by the "counted_by" attribute.
     return nullptr;
 
+  if (isa<DeclRefExpr>(Base))
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/CGBuiltin.cpp:1016]: I'm not sure this reliably detects a "whole struct"?  It's not clear to me why we want to treat `int f(struct X*p) { return __builtin_dynamic_object_size(p, 0); }` differently from `int f2(struct X**p) { return __builtin_dynamic_object_size(*p, 0); }` here.

Maybe this is fine if other parts of the code detect "whole struct" the same way, though.

**Review by efriedma-quic — APPROVED** [review]
LGTM

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 2 above)
Comment (bwendling) [reply clang/lib/CodeGen/CGBuiltin.cpp:1016]: It's possibly a bug, but we currently don't support `int f2(struct X **p) { return __builtin_dynamic_object_size(*p, 0); }` and just return `-1`. We do, however, support something like `&p->array[0]`, which the hope is that it's more common than the previous. From what I can tell, this builtin is kind of a "best effort", so falling back to `-1` when faced with something unexpected should hopefully only result in a bug report.

**Comment by tavianator** [comment]
For the record, I think the most correct definition, in terms of "this is how much memory you should allocate for a struct with a flexible array member" is this:

```c
max(
    sizeof(struct S),      // always at least the size of the struct itself
    round_up(
        alignof(struct S), // size must be a multiple of alignment
        offsetof(struct S, fam) + count * sizeof(((struct S *)0)->fam[0])
    )
)
```

**Comment by Cydox** [comment]
> For the record, I think the most correct definition, in terms of "this is how much memory you should allocate for a struct with a flexible array member" is this:
> 
> ```c
> max(
>     sizeof(struct S),      // always at least the size of the struct itself
>     round_up(
>         alignof(struct S), // size must be a multiple of alignment
>         offsetof(struct S, fam) + count * sizeof(((struct S *)0)->fam[0])
>     )
> )
> ```

I mean it would be useful to round up to the alignment for when you wanne have an array of the structs, but I'm not sure this is actually required by the standard. Do you have more justification for the alignment requirement on structs containing FAMs?

I would argue the minimum "legal" amount to allocate is:
```C
max(
    sizeof(struct S),
    offsetof(struct S, fam) + count * sizeof(((struct S *)0)->fam[0])
)
```

**Comment by Cydox** [comment]
Also see previous discussion here: https://github.com/llvm/llvm-project/pull/111015

**Comment by tavianator** [comment]
> I mean it would be useful to round up to the alignment for when you wanne have an array of the structs, but I'm not sure this is actually required by the standard. Do you have more justification for the alignment requirement on structs containing FAMs?

Here's an example:

```c
struct S {
    int foo;
    char fam[];
};
struct S *s = malloc(9);
s->fam[4];
```

[Your C standard quote](https://github.com/llvm/llvm-project/pull/111015#issuecomment-2392534292) says

> when a . (or ->) operator has a left operand that is (a pointer to) a structure with a flexible array member and the right operand names that member, it behaves as if that member were replaced with the longest array (with the same element type) that would not make the structure larger than the object being accessed

which I interpret as the largest N such that `sizeof(struct S) <= 9` in

```
struct S {
    int foo;
    char fam[N];
};
```

Well, for N = 4 we have `sizeof(struct S) == 8` and for N = 5 we have `sizeof(struct S) == 12` (due to alignment padding), therefore N = 4.  That makes `s->fam[4]` out-of-bounds.  Am I wrong?

**Comment by Cydox** [comment]
I think you're correct

**Comment by Cydox** [comment]
I was forgetting about the padding when thinking about this part of the quote:
> it behaves as if that member were replaced with the longest array (with the same element type) that would not make **the structure** larger than the object being accessed

**Comment by Cydox** [comment]
@bwendling What do you think about that?

**Comment by Cydox** [comment]
The expression can be simplified to:

```C
round_up(
    alignof(struct S)
    offsetof(struct S, fam) + count * sizeof(((struct S *)0)->fam[0])
)
```
As the padding at the end of the structure is always smaller than the alignof.
So `round_up(alignof(struct S), offsetof(struct S, fam)) = sizeof(struct S)`

**Comment by bwendling** [comment]
I think the calculation in https://github.com/llvm/llvm-project/pull/112636#issuecomment-2436559387 is correct. It's probably worthwhile to perform some analysis to see if using that to calculate the new size results in allocation size changes in the kernel. If not, then perhaps we could reinstate the "whole struct" code.

**Comment by Cydox** [comment]
Example of both cases: https://godbolt.org/z/dhzG69sab

**Comment by Cydox** [comment]
We do have to consider though that when `__bdos` is for one of the maximum types (`type & 2 == 0`), it should actually return the largest allowed object that is consistent with the count.

https://github.com/llvm/llvm-project/blob/3b88805ca20018ae202afd3aea39f4fa856a8c64/clang/docs/LanguageExtensions.rst?plain=1#L5502-L5507

So I think the correct result for `type & 2 == 0` is actually:

```C
round_up(
    alignof(struct S)
    offsetof(struct S, fam) + count * sizeof(((struct S *)0)->fam[0])
)
+ alignof(struct S) - 1
```

because all objects that are 1 byte smaller than the calculation in https://github.com/llvm/llvm-project/pull/112636#issuecomment-2436559387 plus `alignof(struct S)` are perfectly legal to have the same count.

Using that calculation for the two examples above looks to actually solve the issue: https://godbolt.org/z/Pdx7Mbano
But that's just after having a quick look at it, we gotta prove that this will actually give a result >= `struct_size` in the kernel in all cases.

**Comment by Cydox** [comment]
> ```
> struct S {
>     int foo;
>     char fam[N];
> };
> ```
> 
> Well, for N = 4 we have `sizeof(struct S) == 8` and for N = 5 we have `sizeof(struct S) == 12` (due to alignment padding), therefore N = 4. That makes `s->fam[4]` out-of-bounds. Am I wrong?

Using this example, every object of sizes 8 to 11 bytes, inclusive behaves like the struct defined with N = 4. Therefore `__bdos(ptr, 0)`, where `ptr` is a pointer to a `struct S` with a `__counted_by` attribute indicating N = 4, should return 11. `__bdos(ptr, 2)` should return 8.

**Comment by Cydox** [comment]
But for now, I think we should merge #112786 into 19.1.3 so that the we can have that be the cutoff in the kernel.

**Comment by tavianator** [comment]
> The expression can be simplified to ... As the padding at the end of the structure is always smaller than the alignof. So `round_up(alignof(struct S), offsetof(struct S, fam)) = sizeof(struct S)`

I think that's right for sane ABIs.  I assume Clang always lays structs out like this.

My use of `max(sizeof(struct S))` was to support pathological compilers that add unnecessary padding.  But it doesn't completely handle pathological compilers, since they may also pad

```c
struct S {
    int foo;
    char fam[N];
};
```

unnecessarily and you can't really predict that.  Ignoring that case, your simplification seems fine to me.

**Pull request merged**